### PR TITLE
feat(PhpdocReadonlyCommentToKeywordFixer): Introduction

### PR DESCRIPTION
**Pull request by EldoranDev**
This MR introduces a new Fixer that will transform the read-only annotations to use the PHP read-only keyword.

should partially fix: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6640

### CODE REVIEW

File: src/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixer.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+/**
+ * @author Marcel Behrmann <marcel@behrmann.dev>
+ */
+final class PhpdocReadonlyCommentToKeywordFixer extends AbstractFixer
+{
+    /**
+     * {@inheritdoc}
+     *
+     * Must run before NoExtraBlankLinesFixer, NoTrailingWhitespaceFixer, NoWhitespaceInBlankLineFixer, PhpdocAlignFixer.
```
Comment (keradus) [inline src/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixer.php:33]: if that's the case, please add tests to `tests/Fixtures/Integration/priority/` folder

File: src/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixer.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            return false;
+        }
+
+        return $tokens->isTokenKindFound(T_DOC_COMMENT);
+    }
+
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition(
+            'Converts readonly comment to readonly keyword.',
```
Comment (keradus) [inline src/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixer.php:52]: does it convert to keyword for everything, or only for classes? if only later, it's OK - but we should mention that in description (and in classname)

File: tests/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixerTest.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    /**
+     * @dataProvider provideFixCases
+     */
+    public function testFix(string $expected, ?string $input = null): void
+    {
+        $this->doTest($expected, $input);
+    }
+
+    public static function provideFixCases(): iterable
+    {
```
Comment (keradus) [inline tests/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixerTest.php:37]: please add this test case

```suggestion
    {
        yield [
            <<<'EOT'
                <?php

                /** Class A. */
                class A {
                }
                EOT,
        ];

```

File: src/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixer.php (hunk 1 above)
Comment (EldoranDev) [reply src/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixer.php]: Added those tests and updated the graph in the FixerFactoryTest.

File: src/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixer.php (hunk 2 above)
Comment (EldoranDev) [reply src/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixer.php]: Updated and moved it into the ClassNotation Namespace.

File: tests/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixerTest.php (hunk 3 above)
Comment (EldoranDev) [reply tests/Fixer/Phpdoc/PhpdocReadonlyCommentToKeywordFixerTest.php]: Added and fixed the test, good catch.

File: src/Fixer/ClassNotation/PhpdocReadonlyClassCommentToKeywordFixer.php [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+            $doc = new DocBlock($token->getContent());
+
+            $annotations = $doc->getAnnotationsOfType('readonly');
+
+            foreach ($annotations as $annotation) {
+                $annotation->remove();
+            }
+
+            if (0 === \count($annotations)) {
+                continue;
```
Comment (keradus) [inline src/Fixer/ClassNotation/PhpdocReadonlyClassCommentToKeywordFixer.php:83]: you can do this early return even before calling foreach loop with `->remove`

File: src/Fixer/ClassNotation/PhpdocReadonlyClassCommentToKeywordFixer.php [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            }
+
+            if ($addReadonly) {
+                $tokens->insertAt($index, [new Token([T_READONLY, 'readonly']), new Token([T_WHITESPACE, ' '])]);
+            }
+
+            $newContent = $doc->getContent();
+
+            if ($newContent === $token->getContent()) {
+                continue;
```
Comment (keradus) [inline src/Fixer/ClassNotation/PhpdocReadonlyClassCommentToKeywordFixer.php:116]: what are the cases in which newContent will be same as previous content?

File: tests/Fixer/ClassNotation/PhpdocReadonlyClassCommentToKeywordFixerTest.php [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+namespace PhpCsFixer\Tests\Fixer\ClassNotation;
+
+use PhpCsFixer\Tests\Test\AbstractFixerTestCase;
+
+/**
+ * @author Marcel Behrmann <marcel@behrmann.dev>
+ *
+ * @internal
+ *
+ * @covers \PhpCsFixer\Fixer\ClassNotation\PhpdocReadonlyClassCommentToKeywordFixer
```
Comment (keradus) [inline tests/Fixer/ClassNotation/PhpdocReadonlyClassCommentToKeywordFixerTest.php:24]: your tests are failing under different PHP versions, please take a look to CI report or run locally using PHP 7.4

File: tests/Fixer/ClassNotation/PhpdocReadonlyClassCommentToKeywordFixerTest.php [hunk 7]
```diff
@@ -27,6 +27,8 @@ final class PhpdocReadonlyClassCommentToKeywordFixerTest extends AbstractFixerTe
 {
     /**
      * @dataProvider provideFixCases
+     *
+     * @requires PHP 8.2
```
Comment (keradus) [inline tests/Fixer/ClassNotation/PhpdocReadonlyClassCommentToKeywordFixerTest.php:31]: FYI - you can also use the annotation for whole class, not only for single test method.

(non blocker, hint for future)

**Review by keradus — APPROVED** [review]
nicely done!

**Comment by Wirone** [comment]
@keradus this fixer should be marked as risky, because it can break user's code (since classes with `@readonly` annotation can be extended anyway, adding keyword would break these child classes).

**Comment by keradus** [comment]
very good point, unfortunate we didn't noticed it earlier in the review.
would you like to create a PR?

**Pull request merged**